File: Novels/packages/5-mobile-ios/StoryReader/ViewModel/CommandPaletteViewModel.swift
import Foundation
import Combine
import SwiftUI

// Command Palette ViewModel - Command palette (mobile-optimized)
class CommandPaletteViewModel: ObservableObject {
    @Published var isOpen: Bool = false
    @Published var searchQuery: String = ""
    @Published var searchResults: [CommandResult] = []
    @Published var isLoading: Bool = false
    
    private let repository = CommandPaletteRepository()
    private var cancellables = Set<AnyCancellable>()
    
    struct CommandResult: Identifiable {
        let id: String
        let type: ResultType
        let title: String
        let subtitle: String?
        let icon: String?
        let action: () -> Void
    }
    
    enum ResultType {
        case story
        case chapter
        case annotation
        case setting
        case action
    }
    
    func search(query: String) {
        searchQuery = query
        isLoading = true
        
        repository.search(query: query)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    self?.isLoading = false
                },
                receiveValue: { [weak self] results in
                    self?.searchResults = results
                }
            )
            .store(in: &cancellables)
    }
    
    func open() {
        isOpen = true
        searchQuery = ""
        searchResults = []
    }
    
    func close() {
        isOpen = false
        searchQuery = ""
        searchResults = []
    }
    
    func executeCommand(_ result: CommandResult) {
        result.action()
        close()
    }
}

File: Novels/packages/6-mobile-android/StoryReader/app/src/main/java/com/storysphere/storyreader/viewmodel/CommandPaletteViewModel.kt
package com.storysphere.storyreader.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class CommandPaletteViewModel @Inject constructor(
    private val storyRepository: com.storysphere.storyreader.repository.StoryRepository,
    private val searchRepository: com.storysphere.storyreader.repository.SearchRepository
) : ViewModel() {
    
    private val _isOpen = MutableStateFlow(false)
    val isOpen: StateFlow<Boolean> = _isOpen.asStateFlow()
    
    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()
    
    private val _searchResults = MutableStateFlow<List<com.storysphere.storyreader.ui.mobile.commandpalette.CommandResult>>(emptyList())
    val searchResults: StateFlow<List<com.storysphere.storyreader.ui.mobile.commandpalette.CommandResult>> = _searchResults.asStateFlow()
    
    fun open() {
        _isOpen.value = true
    }
    
    fun close() {
        _isOpen.value = false
        _searchQuery.value = ""
        _searchResults.value = emptyList()
    }
    
    fun updateSearchQuery(query: String) {
        _searchQuery.value = query
        if (query.isNotEmpty()) {
            performSearch(query)
        } else {
            _searchResults.value = emptyList()
        }
    }
    
    fun search(userId: String, query: String) {
        _searchQuery.value = query
        performSearch(query)
    }
    
    private fun performSearch(query: String) {
        viewModelScope.launch {
            try {
                val results = mutableListOf<com.storysphere.storyreader.ui.mobile.commandpalette.CommandResult>()
                
                // Search stories
                val stories = searchRepository.searchStories(
                    com.storysphere.storyreader.utils.mobile.search.SearchQuery(text = query)
                )
                stories.take(5).forEach { story ->
                    results.add(
                        com.storysphere.storyreader.ui.mobile.commandpalette.CommandResult(
                            type = com.storysphere.storyreader.ui.mobile.commandpalette.ResultType.STORY,
                            id = story.id,
                            title = story.title,
                            subtitle = story.author
                        )
                    )
                }
                
                // Search chapters, annotations, settings could be added here
                
                _searchResults.value = results
            } catch (e: Exception) {
                // Handle error
                _searchResults.value = emptyList()
            }
        }
    }
    
    fun executeCommand(item: CommandPaletteItem) {
        viewModelScope.launch {
            try {
                // Execute command based on item type
                item.action()
                close()
            } catch (e: Exception) {
                // Handle error
            }
        }
    }
}

data class CommandPaletteItem(
    val id: String,
    val title: String,
    val subtitle: String? = null,
    val type: CommandType,
    val action: () -> Unit
)

enum class CommandType {
    STORY,
    CHAPTER,
    ANNOTATION,
    SETTING,
    ACTION
}


